packages/react-json-tree/src/JSONNestedNode.tsx

```diff
@@ -97,144 +97,116 @@ interface Props extends CircularPropsPassedThroughJSONNestedNode {
   expandable: boolean;
 }
 
-interface State {
-  expanded: boolean;
-}
+function JSONNestedNode(props: Props) {
+  const {
+    getItemString,
+    nodeTypeIndicator,
+    nodeType,
+    data = [],
+    hideRoot,
+    createItemString,
+    styling,
+    collectionLimit,
+    keyPath,
+    labelRenderer,
+    expandable = true,
+    isCircular,
+    level = 0,
+    shouldExpandNode,
+  } = props;
 
-function getStateFromProps(props: Props) {
-  // calculate individual node expansion if necessary
-  const expanded = !props.isCircular
-    ? props.shouldExpandNode(props.keyPath, props.data, props.level)
-    : false;
-  return {
-    expanded,
-  };
-}
+  // fixme - call shouldExpandNode to figure out if the should be expanded to start
+  const [expanded, setExpanded] = useState<boolean>(() => {
+    return !isCircular ? shouldExpandNode(keyPath, data, level) : false;
+  });
 
-export default class JSONNestedNode extends React.Component<Props, State> {
-  static propTypes = {
-    getItemString: PropTypes.func.isRequired,
-    nodeTypeIndicator: PropTypes.any,
-    nodeType: PropTypes.string.isRequired,
-    data: PropTypes.any,
-    hideRoot: PropTypes.bool.isRequired,
-    createItemString: PropTypes.func.isRequired,
-    styling: PropTypes.func.isRequired,
-    collectionLimit: PropTypes.number,
-    keyPath: PropTypes.arrayOf(
-      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
-    ).isRequired,
-    labelRenderer: PropTypes.func.isRequired,
-    shouldExpandNode: PropTypes.func,
-    level: PropTypes.number.isRequired,
-    sortObjectKeys: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
-    isCircular: PropTypes.bool,
-    expandable: PropTypes.bool,
-  };
-
-  static defaultProps = {
-    data: [],
-    circularCache: [],
-    level: 0,
-    expandable: true,
-  };
-
-  constructor(props: Props) {
-    super(props);
-    this.state = getStateFromProps(props);
-  }
-
-  UNSAFE_componentWillReceiveProps(nextProps: Props) {
-    const nextState = getStateFromProps(nextProps);
-    if (getStateFromProps(this.props).expanded !== nextState.expanded) {
-      this.setState(nextState);
+  // When certain props change, we need to re-compute whether our node should be in an expanded state
+  useEffect(() => {
+    setExpanded(() => {
+      return !isCircular ? shouldExpandNode(keyPath, data, level) : false;
+    });
+  }, [isCircular, data, keyPath, level, shouldExpandNode]);
+
+  // fixme - previously this was happening after a component should update
```

Comment:
@Methuselah96 Are you familiar with this renderChildNodes code? I believe the intent is that it doesn't run on ever render but only when something has changed. I'm guessing not when all props have changed but only some specific props?

Reply:
It's a little bit complicated because the components can be rendered recursively. For example:

JSONTree -> JSONNode -> JSONObjectNode -> JSONNestedNode -> renderChildNodes -> JSONNode -> ...

That means that `renderChildNodes` does need to run for the majority of prop changes because most of the props get passed down the hierarchy recursively. You could try to create a list of only the props that are recursively passed through the children based on the [circular props type](https://github.com/reduxjs/redux-devtools/blob/c379a6208119aaef5629bddfb3cb1aa447c6a1f4/packages/react-json-tree/src/types.ts#L68-L71), however I would lean toward just calling `renderChildNodes` on every render.

Reply:
👍 Removed the fixme comment - this code now runs on every render